doc/reference/kernel/timing/clocks.rst

```diff
@@ -1,161 +1,242 @@
+which the kernel receives the timeout value).
+
+Other options for timeout initialization follow the unit conventions
+described above: ``K_TIMEOUT_US()``, ``K_TIMEOUT_TICKS()`` and
+``K_TIMEOUT_CYC()`` specify timeout values that will expire after
+specified numbers of microseconds, ticks and cycles, respectively.
+
+Precision of ``k_timeout_t`` values is configurable, with the default
+being 32 bits.  Large uptime counts in non-tick units will experience
+complicated rollover semantics, so it is expected that
+timing-sensitive applications with long uptimes will be configured to
+use a 64 bit timeout type.
+
+Finally, it is possible to specify timeouts as absolute times since
+system boot.  A timeout initialized with ``K_TIMEOUT_UPTIME_MS()``
+indicates a timeout that will expire after the system uptime reaches
+the specified value.  There are likewise microsecond, cycles and ticks
+variants of this API.
+
+Timing Internals
+================
+
+Timeout Queue
+-------------
+
+All Zephyr ``k_timeout_t`` events specified using the API above are
+managed in a single, global queue of events.  Each event is stored in
+a double-linked list, with an attendant delta count in ticks from the
+previous event.  The action to take on an event is specified as a
+callback function pointer provided by the subsystem requesting the
+event, along with a ``struct _timeout`` tracking struct that is
+expected to be embedded within subsystem-defined data structures (for
+example: a ``struct wait_q``, or a ``k_tid_t`` thread struct).
+
+Note that all variant units passed via a ``k_timeout_t`` are converted
+to ticks once on insertion into the list.  There no
+multiple-conversion steps internal to the kernel, so precision is
+guaranteed at the tick level no matter how many events exist or how
+long a timeout might be.
+
+Note that the list structure means that the CPU work involved in
+managing large numbers of timeouts is quadratic in the number of
+active timeouts.  The API design of the timeout queue was intended to
+permit a more scalable backend data structure, but no such
+implementation exists currently.
+
+Timer Drivers
+-------------
+
+Kernel timing at the tick level is driven by a timer driver with a
+comparatively simple API.
+
+* The driver is expected to be able to "announce" new ticks to the
+  kernel via the ``z_clock_nanounce()`` call, which passes an integer
+  number of ticks that have elapsed since the last announce call (or
+  system boot).  These calls can occur at any time, but the driver is
+  expected to attempt to ensure (to the extent practical given
+  interrupt latency interactions) that they occur near tick boundaries
+  (i.e. not "halfway through" a tick), and most importantly that they
+  be correct over time and subject to minimal skew vs. other counters
+  and real world time.
+
+* The driver is expected to provide a ``z_clock_set_timeout()`` call
+  to the kernel which indicates how many ticks may elapse before the
```

Comment:
It may be nice to put a bit more of clear emphasis first in what is expected the function will do. And then, clearly demarked, after, what it should not do.